// igniteui-react-core/esm5/lib/ColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ColumnDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ColumnDescription, _super);
    function ColumnDescription() {
        var _this = _super.call(this) || this;
        _this.da = 0;
        _this.dc = 0;
        _this.db = 0;
        _this.c9 = 0;
        _this.d3 = null;
        _this.dy = null;
        _this.d0 = null;
        _this.dl = null;
        _this.d7 = null;
        _this.dq = null;
        _this.d5 = null;
        _this.dn = null;
        _this.d6 = null;
        _this.dp = null;
        _this.ca = null;
        _this.b6 = null;
        _this.c4 = 0;
        _this.cm = false;
        _this.ch = false;
        _this.b4 = null;
        _this.b8 = null;
        _this.cc = null;
        _this.cl = false;
        _this.co = false;
        _this.cn = false;
        _this.cp = false;
        _this.d8 = null;
        _this.d4 = null;
        _this.dh = null;
        _this.dr = null;
        _this.dj = null;
        _this.dt = null;
        _this.ds = null;
        _this.di = null;
        _this.cf = false;
        _this.ci = false;
        _this.cg = false;
        _this.cj = false;
        _this.ck = false;
        _this.du = null;
        _this.dk = null;
        _this.c3 = 0;
        _this.c1 = 0;
        _this.dv = null;
        _this.c2 = 0;
        _this.dw = null;
        _this.dx = null;
        _this.d2 = null;
        _this.d1 = null;
        _this.dz = null;
        _this.b2 = null;
        _this.dm = null;
        return _this;
    }
    ColumnDescription.prototype.get_type = function () {
        return "Column";
    };
    Object.defineProperty(ColumnDescription.prototype, "paddingLeft", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.e("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "paddingTop", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.e("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "paddingRight", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.e("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "paddingBottom", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.e("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "name", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.e("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "field", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.e("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "headerText", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.e("HeaderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualHeaderText", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.e("ActualHeaderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "selectedBackground", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.e("SelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualSelectedBackground", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.e("ActualSelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "rowHoverBackground", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.e("RowHoverBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualHoverBackground", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.e("ActualHoverBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "rowHoverTextColor", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.e("RowHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualRowHoverTextColor", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.e("ActualRowHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "animationSettings", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.e("AnimationSettings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "width", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.e("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "minWidth", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.e("MinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isFromMarkup", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.e("IsFromMarkup");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isAutoGenerated", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.e("IsAutoGenerated");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "filter", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.e("Filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "filterExpression", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.e("FilterExpression");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "header", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.e("Header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isFilteringEnabled", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.e("IsFilteringEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isResizingEnabled", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.e("IsResizingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isHidden", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.e("IsHidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "shouldRemoveWhenHidden", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.e("ShouldRemoveWhenHidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "sortDirection", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.e("SortDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "pinned", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.e("Pinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualColumnOptionsIconAlignment", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.e("ActualColumnOptionsIconAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "columnOptionsIconAlignment", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.e("ColumnOptionsIconAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualColumnOptionsIconColor", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.e("ActualColumnOptionsIconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "columnOptionsIconColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.e("ColumnOptionsIconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "columnOptionsIconBehavior", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.e("ColumnOptionsIconBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualColumnOptionsIconBehavior", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.e("ActualColumnOptionsIconBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualIsColumnOptionsEnabled", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.e("ActualIsColumnOptionsEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isColumnOptionsEnabled", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.e("IsColumnOptionsEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualIsColumnOptionsSummariesEnabled", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.e("ActualIsColumnOptionsSummariesEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isColumnOptionsSummariesEnabled", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.e("IsColumnOptionsSummariesEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "isEditable", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.e("IsEditable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "deletedTextColor", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.e("DeletedTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualDeletedTextColor", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.e("ActualDeletedTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "editOpacity", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.e("EditOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualEditOpacity", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.e("ActualEditOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "editFontFamily", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.e("EditFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "editFontSize", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.e("EditFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "editFontStyle", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.e("EditFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "editFontWeight", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.e("EditFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "mergedCellMode", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.e("MergedCellMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "mergedCellEvaluationCriteria", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.e("MergedCellEvaluationCriteria");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "filterComparisonType", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.e("FilterComparisonType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "filterOperands", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.e("FilterOperands");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnDescription.prototype, "actualHeaderTextChangeRef", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.e("ActualHeaderTextChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    ColumnDescription.$t = markType(ColumnDescription, 'ColumnDescription', DefinitionBaseDescription.$);
    ColumnDescription.__canMarshalByValue = true;
    return ColumnDescription;
}(DefinitionBaseDescription));
export { ColumnDescription };
