// igniteui-react-core/esm5/lib/WebButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebButtonBaseDescription } from "./WebButtonBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebButtonDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebButtonDescription, _super);
    function WebButtonDescription() {
        var _this = _super.call(this) || this;
        _this.ae = null;
        _this.z = null;
        _this.y = null;
        _this.aa = null;
        return _this;
    }
    WebButtonDescription.prototype.get_type = function () {
        return "WebButton";
    };
    Object.defineProperty(WebButtonDescription.prototype, "variant", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.e("Variant");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonDescription.prototype, "prefix", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.e("Prefix");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonDescription.prototype, "content", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.e("Content");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonDescription.prototype, "suffix", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.e("Suffix");
        },
        enumerable: false,
        configurable: true
    });
    WebButtonDescription.$t = markType(WebButtonDescription, 'WebButtonDescription', WebButtonBaseDescription.$);
    return WebButtonDescription;
}(WebButtonBaseDescription));
export { WebButtonDescription };
