// igniteui-react-core/esm5/lib/WebCardHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCardHeaderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCardHeaderDescription, _super);
    function WebCardHeaderDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.i = null;
        _this.g = null;
        _this.f = null;
        return _this;
    }
    WebCardHeaderDescription.prototype.get_type = function () {
        return "WebCardHeader";
    };
    Object.defineProperty(WebCardHeaderDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCardHeaderDescription.prototype, "thumbnail", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("Thumbnail");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCardHeaderDescription.prototype, "title", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCardHeaderDescription.prototype, "subtitle", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("Subtitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCardHeaderDescription.prototype, "content", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Content");
        },
        enumerable: false,
        configurable: true
    });
    WebCardHeaderDescription.$t = markType(WebCardHeaderDescription, 'WebCardHeaderDescription', Description.$);
    return WebCardHeaderDescription;
}(Description));
export { WebCardHeaderDescription };
