// igniteui-react-core/esm5/lib/WebListItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebListItemDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebListItemDescription, _super);
    function WebListItemDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.i = null;
        _this.f = null;
        _this.g = null;
        return _this;
    }
    WebListItemDescription.prototype.get_type = function () {
        return "WebListItem";
    };
    Object.defineProperty(WebListItemDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebListItemDescription.prototype, "start", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("Start");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebListItemDescription.prototype, "title", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebListItemDescription.prototype, "subtitle", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Subtitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebListItemDescription.prototype, "content", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Content");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebListItemDescription.prototype, "end", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("End");
        },
        enumerable: false,
        configurable: true
    });
    WebListItemDescription.$t = markType(WebListItemDescription, 'WebListItemDescription', Description.$);
    return WebListItemDescription;
}(Description));
export { WebListItemDescription };
