// igniteui-react-core/esm5/lib/WebNavDrawerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebNavDrawerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebNavDrawerDescription, _super);
    function WebNavDrawerDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.j = false;
        _this.f = null;
        _this.g = null;
        _this.o = null;
        _this.n = null;
        _this.m = null;
        _this.l = null;
        return _this;
    }
    WebNavDrawerDescription.prototype.get_type = function () {
        return "WebNavDrawer";
    };
    Object.defineProperty(WebNavDrawerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "position", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("Position");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "open", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "content", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Content");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "mini", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("Mini");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "openingRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("OpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "openedRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("OpenedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "closingRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("ClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebNavDrawerDescription.prototype, "closedRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("ClosedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebNavDrawerDescription.$t = markType(WebNavDrawerDescription, 'WebNavDrawerDescription', Description.$);
    return WebNavDrawerDescription;
}(Description));
export { WebNavDrawerDescription };
