// igniteui-react-core/esm2015/lib/AxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let AxisDescription = /*@__PURE__*/ (() => {
    class AxisDescription extends Description {
        constructor() {
            super();
            this.bm = null;
            this.b4 = null;
            this.b2 = null;
            this.bj = null;
            this.ak = 0;
            this.b1 = null;
            this.bx = null;
            this.bh = null;
            this.ai = 0;
            this.by = null;
            this.bi = null;
            this.aj = 0;
            this.b3 = null;
            this.am = 0;
            this.al = 0;
            this.l = false;
            this.m = false;
            this.o = false;
            this.p = false;
            this.r = false;
            this.bd = 0;
            this.j = false;
            this.bu = null;
            this.bt = null;
            this.n = false;
            this.ab = 0;
            this.ad = 0;
            this.af = 0;
            this.ag = 0;
            this.ah = 0;
            this.ac = 0;
            this.bs = null;
            this.bv = null;
            this.bl = null;
            this.bf = null;
            this.ca = null;
            this.cb = null;
            this.b9 = null;
            this.q = false;
            this.an = 0;
            this.ap = 0;
            this.ar = 0;
            this.as = 0;
            this.at = 0;
            this.ao = 0;
            this.b8 = null;
            this.cc = null;
            this.f = null;
            this.bk = null;
            this.k = false;
            this.br = null;
            this.g = null;
            this.bz = null;
            this.b0 = null;
            this.bn = null;
            this.bo = null;
            this.ae = 0;
            this.bq = null;
            this.bp = null;
            this.b5 = null;
            this.aq = 0;
            this.b7 = null;
            this.b6 = null;
            this.bw = null;
        }
        get_type() {
            return "Axis";
        }
        get type() {
            return this.get_type();
        }
        get formatLabelRef() {
            return this.bm;
        }
        set formatLabelRef(a) {
            this.bm = a;
            this.e("FormatLabelRef");
        }
        get title() {
            return this.b4;
        }
        set title(a) {
            this.b4 = a;
            this.e("Title");
        }
        get stroke() {
            return this.b2;
        }
        set stroke(a) {
            this.b2 = a;
            this.e("Stroke");
        }
        get actualStroke() {
            return this.bj;
        }
        set actualStroke(a) {
            this.bj = a;
            this.e("ActualStroke");
        }
        get strokeThickness() {
            return this.ak;
        }
        set strokeThickness(a) {
            this.ak = a;
            this.e("StrokeThickness");
        }
        get strip() {
            return this.b1;
        }
        set strip(a) {
            this.b1 = a;
            this.e("Strip");
        }
        get majorStroke() {
            return this.bx;
        }
        set majorStroke(a) {
            this.bx = a;
            this.e("MajorStroke");
        }
        get actualMajorStroke() {
            return this.bh;
        }
        set actualMajorStroke(a) {
            this.bh = a;
            this.e("ActualMajorStroke");
        }
        get majorStrokeThickness() {
            return this.ai;
        }
        set majorStrokeThickness(a) {
            this.ai = a;
            this.e("MajorStrokeThickness");
        }
        get minorStroke() {
            return this.by;
        }
        set minorStroke(a) {
            this.by = a;
            this.e("MinorStroke");
        }
        get actualMinorStroke() {
            return this.bi;
        }
        set actualMinorStroke(a) {
            this.bi = a;
            this.e("ActualMinorStroke");
        }
        get minorStrokeThickness() {
            return this.aj;
        }
        set minorStrokeThickness(a) {
            this.aj = a;
            this.e("MinorStrokeThickness");
        }
        get tickStroke() {
            return this.b3;
        }
        set tickStroke(a) {
            this.b3 = a;
            this.e("TickStroke");
        }
        get tickStrokeThickness() {
            return this.am;
        }
        set tickStrokeThickness(a) {
            this.am = a;
            this.e("TickStrokeThickness");
        }
        get tickLength() {
            return this.al;
        }
        set tickLength(a) {
            this.al = a;
            this.e("TickLength");
        }
        get isDisabled() {
            return this.l;
        }
        set isDisabled(a) {
            this.l = a;
            this.e("IsDisabled");
        }
        get isInverted() {
            return this.m;
        }
        set isInverted(a) {
            this.m = a;
            this.e("IsInverted");
        }
        get shouldAvoidAnnotationCollisions() {
            return this.o;
        }
        set shouldAvoidAnnotationCollisions(a) {
            this.o = a;
            this.e("ShouldAvoidAnnotationCollisions");
        }
        get shouldKeepAnnotationsInView() {
            return this.p;
        }
        set shouldKeepAnnotationsInView(a) {
            this.p = a;
            this.e("ShouldKeepAnnotationsInView");
        }
        get useEnhancedIntervalManagement() {
            return this.r;
        }
        set useEnhancedIntervalManagement(a) {
            this.r = a;
            this.e("UseEnhancedIntervalManagement");
        }
        get enhancedIntervalMinimumCharacters() {
            return this.bd;
        }
        set enhancedIntervalMinimumCharacters(a) {
            this.bd = a;
            this.e("EnhancedIntervalMinimumCharacters");
        }
        get enhancedIntervalPreferMoreCategoryLabels() {
            return this.j;
        }
        set enhancedIntervalPreferMoreCategoryLabels(a) {
            this.j = a;
            this.e("EnhancedIntervalPreferMoreCategoryLabels");
        }
        get labelTextColor() {
            return this.bu;
        }
        set labelTextColor(a) {
            this.bu = a;
            this.e("LabelTextColor");
        }
        get labelLocation() {
            return this.bt;
        }
        set labelLocation(a) {
            this.bt = a;
            this.e("LabelLocation");
        }
        get labelShowFirstLabel() {
            return this.n;
        }
        set labelShowFirstLabel(a) {
            this.n = a;
            this.e("LabelShowFirstLabel");
        }
        get labelAngle() {
            return this.ab;
        }
        set labelAngle(a) {
            this.ab = a;
            this.e("LabelAngle");
        }
        get labelExtent() {
            return this.ad;
        }
        set labelExtent(a) {
            this.ad = a;
            this.e("LabelExtent");
        }
        get labelLeftMargin() {
            return this.af;
        }
        set labelLeftMargin(a) {
            this.af = a;
            this.e("LabelLeftMargin");
        }
        get labelRightMargin() {
            return this.ag;
        }
        set labelRightMargin(a) {
            this.ag = a;
            this.e("LabelRightMargin");
        }
        get labelTopMargin() {
            return this.ah;
        }
        set labelTopMargin(a) {
            this.ah = a;
            this.e("LabelTopMargin");
        }
        get labelBottomMargin() {
            return this.ac;
        }
        set labelBottomMargin(a) {
            this.ac = a;
            this.e("LabelBottomMargin");
        }
        get labelHorizontalAlignment() {
            return this.bs;
        }
        set labelHorizontalAlignment(a) {
            this.bs = a;
            this.e("LabelHorizontalAlignment");
        }
        get labelVerticalAlignment() {
            return this.bv;
        }
        set labelVerticalAlignment(a) {
            this.bv = a;
            this.e("LabelVerticalAlignment");
        }
        get crossingAxisRef() {
            return this.bl;
        }
        set crossingAxisRef(a) {
            this.bl = a;
            this.e("CrossingAxisRef");
        }
        get crossingValue() {
            return this.bf;
        }
        set crossingValue(a) {
            this.bf = a;
            this.e("CrossingValue");
        }
        get titlePosition() {
            return this.ca;
        }
        set titlePosition(a) {
            this.ca = a;
            this.e("TitlePosition");
        }
        get titleTextColor() {
            return this.cb;
        }
        set titleTextColor(a) {
            this.cb = a;
            this.e("TitleTextColor");
        }
        get titleLocation() {
            return this.b9;
        }
        set titleLocation(a) {
            this.b9 = a;
            this.e("TitleLocation");
        }
        get titleShowFirstLabel() {
            return this.q;
        }
        set titleShowFirstLabel(a) {
            this.q = a;
            this.e("TitleShowFirstLabel");
        }
        get titleAngle() {
            return this.an;
        }
        set titleAngle(a) {
            this.an = a;
            this.e("TitleAngle");
        }
        get titleExtent() {
            return this.ap;
        }
        set titleExtent(a) {
            this.ap = a;
            this.e("TitleExtent");
        }
        get titleLeftMargin() {
            return this.ar;
        }
        set titleLeftMargin(a) {
            this.ar = a;
            this.e("TitleLeftMargin");
        }
        get titleRightMargin() {
            return this.as;
        }
        set titleRightMargin(a) {
            this.as = a;
            this.e("TitleRightMargin");
        }
        get titleTopMargin() {
            return this.at;
        }
        set titleTopMargin(a) {
            this.at = a;
            this.e("TitleTopMargin");
        }
        get titleBottomMargin() {
            return this.ao;
        }
        set titleBottomMargin(a) {
            this.ao = a;
            this.e("TitleBottomMargin");
        }
        get titleHorizontalAlignment() {
            return this.b8;
        }
        set titleHorizontalAlignment(a) {
            this.b8 = a;
            this.e("TitleHorizontalAlignment");
        }
        get titleVerticalAlignment() {
            return this.cc;
        }
        set titleVerticalAlignment(a) {
            this.cc = a;
            this.e("TitleVerticalAlignment");
        }
        get annotations() {
            return this.f;
        }
        set annotations(a) {
            this.f = a;
            this.e("Annotations");
        }
        get coercionMethodsRef() {
            return this.bk;
        }
        set coercionMethodsRef(a) {
            this.bk = a;
            this.e("CoercionMethodsRef");
        }
        get expectFunctions() {
            return this.k;
        }
        set expectFunctions(a) {
            this.k = a;
            this.e("ExpectFunctions");
        }
        get labelFormat() {
            return this.br;
        }
        set labelFormat(a) {
            this.br = a;
            this.e("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.g;
        }
        set labelFormatSpecifiers(a) {
            this.g = a;
            this.e("LabelFormatSpecifiers");
        }
        get rangeChangedRef() {
            return this.bz;
        }
        set rangeChangedRef(a) {
            this.bz = a;
            this.e("RangeChangedRef");
        }
        get renderRequestedRef() {
            return this.b0;
        }
        set renderRequestedRef(a) {
            this.b0 = a;
            this.e("RenderRequestedRef");
        }
        get label() {
            return this.bn;
        }
        set label(a) {
            this.bn = a;
            this.e("Label");
        }
        get labelFontFamily() {
            return this.bo;
        }
        set labelFontFamily(a) {
            this.bo = a;
            this.e("LabelFontFamily");
        }
        get labelFontSize() {
            return this.ae;
        }
        set labelFontSize(a) {
            this.ae = a;
            this.e("LabelFontSize");
        }
        get labelFontWeight() {
            return this.bq;
        }
        set labelFontWeight(a) {
            this.bq = a;
            this.e("LabelFontWeight");
        }
        get labelFontStyle() {
            return this.bp;
        }
        set labelFontStyle(a) {
            this.bp = a;
            this.e("LabelFontStyle");
        }
        get titleFontFamily() {
            return this.b5;
        }
        set titleFontFamily(a) {
            this.b5 = a;
            this.e("TitleFontFamily");
        }
        get titleFontSize() {
            return this.aq;
        }
        set titleFontSize(a) {
            this.aq = a;
            this.e("TitleFontSize");
        }
        get titleFontWeight() {
            return this.b7;
        }
        set titleFontWeight(a) {
            this.b7 = a;
            this.e("TitleFontWeight");
        }
        get titleFontStyle() {
            return this.b6;
        }
        set titleFontStyle(a) {
            this.b6 = a;
            this.e("TitleFontStyle");
        }
        get labelVisibility() {
            return this.bw;
        }
        set labelVisibility(a) {
            this.bw = a;
            this.e("LabelVisibility");
        }
    }
    AxisDescription.$t = /*@__PURE__*/ markType(AxisDescription, 'AxisDescription', Description.$);
    return AxisDescription;
})();
