// igniteui-react-core/esm2015/lib/CategoryAngleAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAxisBaseDescription } from "./CategoryAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryAngleAxisDescription = /*@__PURE__*/ (() => {
    class CategoryAngleAxisDescription extends CategoryAxisBaseDescription {
        constructor() {
            super();
            this.dr = 0;
            this.dp = 0;
            this.dm = 0;
            this.dq = 0;
            this.dn = 0;
            this.dx = null;
        }
        get_type() {
            return "CategoryAngleAxis";
        }
        get startAngleOffset() {
            return this.dr;
        }
        set startAngleOffset(a) {
            this.dr = a;
            this.e("StartAngleOffset");
        }
        get interval() {
            return this.dp;
        }
        set interval(a) {
            this.dp = a;
            this.e("Interval");
        }
        get actualInterval() {
            return this.dm;
        }
        set actualInterval(a) {
            this.dm = a;
            this.e("ActualInterval");
        }
        get minorInterval() {
            return this.dq;
        }
        set minorInterval(a) {
            this.dq = a;
            this.e("MinorInterval");
        }
        get actualMinorInterval() {
            return this.dn;
        }
        set actualMinorInterval(a) {
            this.dn = a;
            this.e("ActualMinorInterval");
        }
        get actualIntervalChangeRef() {
            return this.dx;
        }
        set actualIntervalChangeRef(a) {
            this.dx = a;
            this.e("ActualIntervalChangeRef");
        }
    }
    CategoryAngleAxisDescription.$t = /*@__PURE__*/ markType(CategoryAngleAxisDescription, 'CategoryAngleAxisDescription', CategoryAxisBaseDescription.$);
    return CategoryAngleAxisDescription;
})();
