// igniteui-react-core/esm2015/lib/CategoryAxisBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AxisDescription } from "./AxisDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryAxisBaseDescription = /*@__PURE__*/ (() => {
    class CategoryAxisBaseDescription extends AxisDescription {
        constructor() {
            super();
            this.di = null;
            this.dg = 0;
            this.dc = 0;
            this.dd = 0;
            this.da = false;
            this.dj = null;
        }
        get_type() {
            return "CategoryAxisBase";
        }
        get dataSourceRef() {
            return this.di;
        }
        set dataSourceRef(a) {
            this.di = a;
            this.e("DataSourceRef");
        }
        get itemsCount() {
            return this.dg;
        }
        set itemsCount(a) {
            this.dg = a;
            this.e("ItemsCount");
        }
        get gap() {
            return this.dc;
        }
        set gap(a) {
            this.dc = a;
            this.e("Gap");
        }
        get overlap() {
            return this.dd;
        }
        set overlap(a) {
            this.dd = a;
            this.e("Overlap");
        }
        get useClusteringMode() {
            return this.da;
        }
        set useClusteringMode(a) {
            this.da = a;
            this.e("UseClusteringMode");
        }
        get itemsCountChangeRef() {
            return this.dj;
        }
        set itemsCountChangeRef(a) {
            this.dj = a;
            this.e("ItemsCountChangeRef");
        }
    }
    CategoryAxisBaseDescription.$t = /*@__PURE__*/ markType(CategoryAxisBaseDescription, 'CategoryAxisBaseDescription', AxisDescription.$);
    return CategoryAxisBaseDescription;
})();
