// igniteui-react-core/esm2015/lib/ColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnDescription = /*@__PURE__*/ (() => {
    class ColumnDescription extends DefinitionBaseDescription {
        constructor() {
            super();
            this.da = 0;
            this.dc = 0;
            this.db = 0;
            this.c9 = 0;
            this.d3 = null;
            this.dy = null;
            this.d0 = null;
            this.dl = null;
            this.d7 = null;
            this.dq = null;
            this.d5 = null;
            this.dn = null;
            this.d6 = null;
            this.dp = null;
            this.ca = null;
            this.b6 = null;
            this.c4 = 0;
            this.cm = false;
            this.ch = false;
            this.b4 = null;
            this.b8 = null;
            this.cc = null;
            this.cl = false;
            this.co = false;
            this.cn = false;
            this.cp = false;
            this.d8 = null;
            this.d4 = null;
            this.dh = null;
            this.dr = null;
            this.dj = null;
            this.dt = null;
            this.ds = null;
            this.di = null;
            this.cf = false;
            this.ci = false;
            this.cg = false;
            this.cj = false;
            this.ck = false;
            this.du = null;
            this.dk = null;
            this.c3 = 0;
            this.c1 = 0;
            this.dv = null;
            this.c2 = 0;
            this.dw = null;
            this.dx = null;
            this.d2 = null;
            this.d1 = null;
            this.dz = null;
            this.b2 = null;
            this.dm = null;
        }
        get_type() {
            return "Column";
        }
        get paddingLeft() {
            return this.da;
        }
        set paddingLeft(a) {
            this.da = a;
            this.e("PaddingLeft");
        }
        get paddingTop() {
            return this.dc;
        }
        set paddingTop(a) {
            this.dc = a;
            this.e("PaddingTop");
        }
        get paddingRight() {
            return this.db;
        }
        set paddingRight(a) {
            this.db = a;
            this.e("PaddingRight");
        }
        get paddingBottom() {
            return this.c9;
        }
        set paddingBottom(a) {
            this.c9 = a;
            this.e("PaddingBottom");
        }
        get name() {
            return this.d3;
        }
        set name(a) {
            this.d3 = a;
            this.e("Name");
        }
        get field() {
            return this.dy;
        }
        set field(a) {
            this.dy = a;
            this.e("Field");
        }
        get headerText() {
            return this.d0;
        }
        set headerText(a) {
            this.d0 = a;
            this.e("HeaderText");
        }
        get actualHeaderText() {
            return this.dl;
        }
        set actualHeaderText(a) {
            this.dl = a;
            this.e("ActualHeaderText");
        }
        get selectedBackground() {
            return this.d7;
        }
        set selectedBackground(a) {
            this.d7 = a;
            this.e("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.dq;
        }
        set actualSelectedBackground(a) {
            this.dq = a;
            this.e("ActualSelectedBackground");
        }
        get rowHoverBackground() {
            return this.d5;
        }
        set rowHoverBackground(a) {
            this.d5 = a;
            this.e("RowHoverBackground");
        }
        get actualHoverBackground() {
            return this.dn;
        }
        set actualHoverBackground(a) {
            this.dn = a;
            this.e("ActualHoverBackground");
        }
        get rowHoverTextColor() {
            return this.d6;
        }
        set rowHoverTextColor(a) {
            this.d6 = a;
            this.e("RowHoverTextColor");
        }
        get actualRowHoverTextColor() {
            return this.dp;
        }
        set actualRowHoverTextColor(a) {
            this.dp = a;
            this.e("ActualRowHoverTextColor");
        }
        get animationSettings() {
            return this.ca;
        }
        set animationSettings(a) {
            this.ca = a;
            this.e("AnimationSettings");
        }
        get width() {
            return this.b6;
        }
        set width(a) {
            this.b6 = a;
            this.e("Width");
        }
        get minWidth() {
            return this.c4;
        }
        set minWidth(a) {
            this.c4 = a;
            this.e("MinWidth");
        }
        get isFromMarkup() {
            return this.cm;
        }
        set isFromMarkup(a) {
            this.cm = a;
            this.e("IsFromMarkup");
        }
        get isAutoGenerated() {
            return this.ch;
        }
        set isAutoGenerated(a) {
            this.ch = a;
            this.e("IsAutoGenerated");
        }
        get filter() {
            return this.b4;
        }
        set filter(a) {
            this.b4 = a;
            this.e("Filter");
        }
        get filterExpression() {
            return this.b8;
        }
        set filterExpression(a) {
            this.b8 = a;
            this.e("FilterExpression");
        }
        get header() {
            return this.cc;
        }
        set header(a) {
            this.cc = a;
            this.e("Header");
        }
        get isFilteringEnabled() {
            return this.cl;
        }
        set isFilteringEnabled(a) {
            this.cl = a;
            this.e("IsFilteringEnabled");
        }
        get isResizingEnabled() {
            return this.co;
        }
        set isResizingEnabled(a) {
            this.co = a;
            this.e("IsResizingEnabled");
        }
        get isHidden() {
            return this.cn;
        }
        set isHidden(a) {
            this.cn = a;
            this.e("IsHidden");
        }
        get shouldRemoveWhenHidden() {
            return this.cp;
        }
        set shouldRemoveWhenHidden(a) {
            this.cp = a;
            this.e("ShouldRemoveWhenHidden");
        }
        get sortDirection() {
            return this.d8;
        }
        set sortDirection(a) {
            this.d8 = a;
            this.e("SortDirection");
        }
        get pinned() {
            return this.d4;
        }
        set pinned(a) {
            this.d4 = a;
            this.e("Pinned");
        }
        get actualColumnOptionsIconAlignment() {
            return this.dh;
        }
        set actualColumnOptionsIconAlignment(a) {
            this.dh = a;
            this.e("ActualColumnOptionsIconAlignment");
        }
        get columnOptionsIconAlignment() {
            return this.dr;
        }
        set columnOptionsIconAlignment(a) {
            this.dr = a;
            this.e("ColumnOptionsIconAlignment");
        }
        get actualColumnOptionsIconColor() {
            return this.dj;
        }
        set actualColumnOptionsIconColor(a) {
            this.dj = a;
            this.e("ActualColumnOptionsIconColor");
        }
        get columnOptionsIconColor() {
            return this.dt;
        }
        set columnOptionsIconColor(a) {
            this.dt = a;
            this.e("ColumnOptionsIconColor");
        }
        get columnOptionsIconBehavior() {
            return this.ds;
        }
        set columnOptionsIconBehavior(a) {
            this.ds = a;
            this.e("ColumnOptionsIconBehavior");
        }
        get actualColumnOptionsIconBehavior() {
            return this.di;
        }
        set actualColumnOptionsIconBehavior(a) {
            this.di = a;
            this.e("ActualColumnOptionsIconBehavior");
        }
        get actualIsColumnOptionsEnabled() {
            return this.cf;
        }
        set actualIsColumnOptionsEnabled(a) {
            this.cf = a;
            this.e("ActualIsColumnOptionsEnabled");
        }
        get isColumnOptionsEnabled() {
            return this.ci;
        }
        set isColumnOptionsEnabled(a) {
            this.ci = a;
            this.e("IsColumnOptionsEnabled");
        }
        get actualIsColumnOptionsSummariesEnabled() {
            return this.cg;
        }
        set actualIsColumnOptionsSummariesEnabled(a) {
            this.cg = a;
            this.e("ActualIsColumnOptionsSummariesEnabled");
        }
        get isColumnOptionsSummariesEnabled() {
            return this.cj;
        }
        set isColumnOptionsSummariesEnabled(a) {
            this.cj = a;
            this.e("IsColumnOptionsSummariesEnabled");
        }
        get isEditable() {
            return this.ck;
        }
        set isEditable(a) {
            this.ck = a;
            this.e("IsEditable");
        }
        get deletedTextColor() {
            return this.du;
        }
        set deletedTextColor(a) {
            this.du = a;
            this.e("DeletedTextColor");
        }
        get actualDeletedTextColor() {
            return this.dk;
        }
        set actualDeletedTextColor(a) {
            this.dk = a;
            this.e("ActualDeletedTextColor");
        }
        get editOpacity() {
            return this.c3;
        }
        set editOpacity(a) {
            this.c3 = a;
            this.e("EditOpacity");
        }
        get actualEditOpacity() {
            return this.c1;
        }
        set actualEditOpacity(a) {
            this.c1 = a;
            this.e("ActualEditOpacity");
        }
        get editFontFamily() {
            return this.dv;
        }
        set editFontFamily(a) {
            this.dv = a;
            this.e("EditFontFamily");
        }
        get editFontSize() {
            return this.c2;
        }
        set editFontSize(a) {
            this.c2 = a;
            this.e("EditFontSize");
        }
        get editFontStyle() {
            return this.dw;
        }
        set editFontStyle(a) {
            this.dw = a;
            this.e("EditFontStyle");
        }
        get editFontWeight() {
            return this.dx;
        }
        set editFontWeight(a) {
            this.dx = a;
            this.e("EditFontWeight");
        }
        get mergedCellMode() {
            return this.d2;
        }
        set mergedCellMode(a) {
            this.d2 = a;
            this.e("MergedCellMode");
        }
        get mergedCellEvaluationCriteria() {
            return this.d1;
        }
        set mergedCellEvaluationCriteria(a) {
            this.d1 = a;
            this.e("MergedCellEvaluationCriteria");
        }
        get filterComparisonType() {
            return this.dz;
        }
        set filterComparisonType(a) {
            this.dz = a;
            this.e("FilterComparisonType");
        }
        get filterOperands() {
            return this.b2;
        }
        set filterOperands(a) {
            this.b2 = a;
            this.e("FilterOperands");
        }
        get actualHeaderTextChangeRef() {
            return this.dm;
        }
        set actualHeaderTextChangeRef(a) {
            this.dm = a;
            this.e("ActualHeaderTextChangeRef");
        }
    }
    ColumnDescription.$t = /*@__PURE__*/ markType(ColumnDescription, 'ColumnDescription', DefinitionBaseDescription.$);
    ColumnDescription.__canMarshalByValue = true;
    return ColumnDescription;
})();
