// igniteui-react-core/esm2015/lib/ColumnGroupingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnGroupingDescription = /*@__PURE__*/ (() => {
    class ColumnGroupingDescription extends Description {
        constructor() {
            super();
            this.p = null;
            this.i = null;
            this.m = null;
            this.r = null;
            this.s = null;
            this.j = null;
            this.f = 0;
            this.k = null;
            this.l = null;
            this.q = null;
            this.o = null;
            this.h = null;
            this.n = null;
        }
        get_type() {
            return "ColumnGrouping";
        }
        get type() {
            return this.get_type();
        }
        get targetGridRef() {
            return this.p;
        }
        set targetGridRef(a) {
            this.p = a;
            this.e("TargetGridRef");
        }
        get baseTheme() {
            return this.i;
        }
        set baseTheme(a) {
            this.i = a;
            this.e("BaseTheme");
        }
        get density() {
            return this.m;
        }
        set density(a) {
            this.m = a;
            this.e("Density");
        }
        get title() {
            return this.r;
        }
        set title(a) {
            this.r = a;
            this.e("Title");
        }
        get titleColor() {
            return this.s;
        }
        set titleColor(a) {
            this.s = a;
            this.e("TitleColor");
        }
        get columnGroupTitleFontFamily() {
            return this.j;
        }
        set columnGroupTitleFontFamily(a) {
            this.j = a;
            this.e("ColumnGroupTitleFontFamily");
        }
        get columnGroupTitleFontSize() {
            return this.f;
        }
        set columnGroupTitleFontSize(a) {
            this.f = a;
            this.e("ColumnGroupTitleFontSize");
        }
        get columnGroupTitleFontStyle() {
            return this.k;
        }
        set columnGroupTitleFontStyle(a) {
            this.k = a;
            this.e("ColumnGroupTitleFontStyle");
        }
        get columnGroupTitleFontWeight() {
            return this.l;
        }
        set columnGroupTitleFontWeight(a) {
            this.l = a;
            this.e("ColumnGroupTitleFontWeight");
        }
        get textColor() {
            return this.q;
        }
        set textColor(a) {
            this.q = a;
            this.e("TextColor");
        }
        get iconColor() {
            return this.o;
        }
        set iconColor(a) {
            this.o = a;
            this.e("IconColor");
        }
        get backgroundColor() {
            return this.h;
        }
        set backgroundColor(a) {
            this.h = a;
            this.e("BackgroundColor");
        }
        get hoverBackgroundColor() {
            return this.n;
        }
        set hoverBackgroundColor(a) {
            this.n = a;
            this.e("HoverBackgroundColor");
        }
    }
    ColumnGroupingDescription.$t = /*@__PURE__*/ markType(ColumnGroupingDescription, 'ColumnGroupingDescription', Description.$);
    return ColumnGroupingDescription;
})();
