// igniteui-react-core/esm2015/lib/ColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnSeriesDescription = /*@__PURE__*/ (() => {
    class ColumnSeriesDescription extends HorizontalAnchoredCategorySeriesDescription {
        constructor() {
            super();
            this.eh = 0;
            this.ei = 0;
            this.el = null;
        }
        get_type() {
            return "ColumnSeries";
        }
        get radiusX() {
            return this.eh;
        }
        set radiusX(a) {
            this.eh = a;
            this.e("RadiusX");
        }
        get radiusY() {
            return this.ei;
        }
        set radiusY(a) {
            this.ei = a;
            this.e("RadiusY");
        }
        get consolidatedColumnVerticalPosition() {
            return this.el;
        }
        set consolidatedColumnVerticalPosition(a) {
            this.el = a;
            this.e("ConsolidatedColumnVerticalPosition");
        }
    }
    ColumnSeriesDescription.$t = /*@__PURE__*/ markType(ColumnSeriesDescription, 'ColumnSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return ColumnSeriesDescription;
})();
