// igniteui-react-core/esm2015/lib/DataGridComparisonOperatorSelectorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridComparisonOperatorSelectorDescription = /*@__PURE__*/ (() => {
    class DataGridComparisonOperatorSelectorDescription extends Description {
        constructor() {
            super();
            this.m = null;
            this.f = 0;
            this.k = null;
            this.g = 0;
            this.n = null;
            this.l = null;
            this.j = null;
        }
        get_type() {
            return "DataGridComparisonOperatorSelector";
        }
        get type() {
            return this.get_type();
        }
        get value() {
            return this.m;
        }
        set value(a) {
            this.m = a;
            this.e("Value");
        }
        get customFilterID() {
            return this.f;
        }
        set customFilterID(a) {
            this.f = a;
            this.e("CustomFilterID");
        }
        get dataType() {
            return this.k;
        }
        set dataType(a) {
            this.k = a;
            this.e("DataType");
        }
        get maxHeight() {
            return this.g;
        }
        set maxHeight(a) {
            this.g = a;
            this.e("MaxHeight");
        }
        get valueChangedRef() {
            return this.n;
        }
        set valueChangedRef(a) {
            this.n = a;
            this.e("ValueChangedRef");
        }
        get openingRef() {
            return this.l;
        }
        set openingRef(a) {
            this.l = a;
            this.e("OpeningRef");
        }
        get closingRef() {
            return this.j;
        }
        set closingRef(a) {
            this.j = a;
            this.e("ClosingRef");
        }
    }
    DataGridComparisonOperatorSelectorDescription.$t = /*@__PURE__*/ markType(DataGridComparisonOperatorSelectorDescription, 'DataGridComparisonOperatorSelectorDescription', Description.$);
    return DataGridComparisonOperatorSelectorDescription;
})();
