// igniteui-react-core/esm2015/lib/EditorCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let EditorCellInfoDescription = /*@__PURE__*/ (() => {
    class EditorCellInfoDescription extends CellInfoDescription {
        constructor() {
            super();
            this.fl = null;
            this.fp = null;
            this.fv = null;
            this.fu = null;
            this.ft = null;
            this.fr = null;
            this.fs = null;
            this.fj = null;
            this.fq = null;
        }
        get_type() {
            return "EditorCellInfo";
        }
        get editTarget() {
            return this.fl;
        }
        set editTarget(a) {
            this.fl = a;
            this.e("EditTarget");
        }
        get dataType() {
            return this.fp;
        }
        set dataType(a) {
            this.fp = a;
            this.e("DataType");
        }
        get errorMessage() {
            return this.fv;
        }
        set errorMessage(a) {
            this.fv = a;
            this.e("ErrorMessage");
        }
        get editValueRef() {
            return this.fu;
        }
        set editValueRef(a) {
            this.fu = a;
            this.e("EditValueRef");
        }
        get editorType() {
            return this.ft;
        }
        set editorType(a) {
            this.ft = a;
            this.e("EditorType");
        }
        get editorDataSourceRef() {
            return this.fr;
        }
        set editorDataSourceRef(a) {
            this.fr = a;
            this.e("EditorDataSourceRef");
        }
        get editorTextField() {
            return this.fs;
        }
        set editorTextField(a) {
            this.fs = a;
            this.e("EditorTextField");
        }
        get editorValueField() {
            return this.fj;
        }
        set editorValueField(a) {
            this.fj = a;
            this.e("EditorValueField");
        }
        get dateFormatString() {
            return this.fq;
        }
        set dateFormatString(a) {
            this.fq = a;
            this.e("DateFormatString");
        }
    }
    EditorCellInfoDescription.$t = /*@__PURE__*/ markType(EditorCellInfoDescription, 'EditorCellInfoDescription', CellInfoDescription.$);
    EditorCellInfoDescription.__marshalByValue1 = true;
    EditorCellInfoDescription.__marshalByValueAlias1 = "EditorCellModel";
    return EditorCellInfoDescription;
})();
