// igniteui-react-core/esm2015/lib/EventProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let EventProxy = /*@__PURE__*/ (() => {
    class EventProxy extends Base {
        constructor() {
            super(...arguments);
            this.onMouseWheel = null;
            this.onMouseWheelHorizontal = null;
            this.onPinchStarted = null;
            this.onPinchDelta = null;
            this.onGestureCompleted = null;
            this.onZoomStarted = null;
            this.onZoomDelta = null;
            this.onZoomCompleted = null;
            this.onFlingStarted = null;
            this.onContactStarted = null;
            this.onDragStarted = null;
            this.onContactMoved = null;
            this.onDragDelta = null;
            this.onContactCompleted = null;
            this.onDragCompleted = null;
            this.onMouseLeave = null;
            this.onMouseOver = null;
            this.onMouseEnter = null;
            this.onMouseDown = null;
            this.onMouseUp = null;
            this.onDoubleTap = null;
            this.onHold = null;
            this.onKeyDown = null;
            this.onKeyUp = null;
            this.bg = null;
            this.bd = 0;
            this.ai = false;
            this.aj = false;
            this.al = null;
            this.an = null;
            this.ao = null;
            this.ap = null;
            this.z = false;
            this.aa = false;
            this.am = null;
        }
        ab(a, b, c) {
            if (this.an != null) {
                return this.an(a, b, c);
            }
            return this.al(a);
        }
        get_ak() {
            return false;
        }
        get ak() {
            return this.get_ak();
        }
        ag(a, b) {
            if (this.onMouseWheel != null && !this.ac) {
                return this.onMouseWheel(a, b);
            }
            return false;
        }
        ah(a, b) {
            if (this.onMouseWheelHorizontal != null && !this.ac) {
                return this.onMouseWheelHorizontal(a, b);
            }
            return false;
        }
        a8(a, b) {
            if (this.onPinchStarted != null && !this.ac) {
                this.onPinchStarted(a, b);
            }
        }
        bb(a, b, c, d) {
            if (this.onZoomStarted != null && !this.ac) {
                this.onZoomStarted(a, b, c, d);
            }
        }
        ad(a, b, c) {
            if (this.onFlingStarted != null && !this.ac) {
                return this.onFlingStarted(a, b, c);
            }
            return true;
        }
        a7(a, b) {
            if (this.onPinchDelta != null && !this.ac) {
                this.onPinchDelta(a, b);
            }
        }
        ba(a, b, c, d) {
            if (this.onZoomDelta != null && !this.ac) {
                this.onZoomDelta(a, b, c, d);
            }
        }
        a9(a, b, c, d) {
            if (this.onZoomCompleted != null && !this.ac) {
                this.onZoomCompleted(a, b, c, d);
            }
        }
        a0(a, b) {
            if (this.onGestureCompleted != null && !this.ac) {
                this.onGestureCompleted(a, b);
            }
        }
        av(a, b) {
            if (this.onContactStarted != null && !this.ac) {
                this.onContactStarted(a, b);
            }
        }
        az(a) {
            if (this.onDragStarted != null && !this.ac) {
                this.onDragStarted(a);
            }
        }
        au(a, b) {
            if (this.onContactMoved != null && !this.ac) {
                this.onContactMoved(a, b);
            }
        }
        ay(a) {
            if (this.onDragDelta != null && !this.ac) {
                this.onDragDelta(a);
            }
        }
        at(a, b, c) {
            if (this.onContactCompleted != null && !this.ac) {
                this.aj = c;
                this.onContactCompleted(a, b);
                this.aj = false;
            }
        }
        ax(a) {
            if (this.onDragCompleted != null && !this.ac) {
                this.onDragCompleted(a);
            }
        }
        a4(a) {
            if (this.onMouseLeave != null && !this.ac) {
                this.onMouseLeave(a);
            }
        }
        a5(a, b, c) {
            if (this.onMouseOver != null && !this.ac) {
                this.onMouseOver(a, b, c);
            }
        }
        a3(a) {
            if (this.onMouseEnter != null && !this.ac) {
                this.onMouseEnter(a);
            }
        }
        a2(a) {
            if (this.onMouseDown != null && !this.ac) {
                this.onMouseDown(a);
            }
        }
        a6(a) {
            if (this.onMouseUp != null && !this.ac) {
                this.onMouseUp(a);
            }
        }
        ae(a) {
            if (this.onKeyDown != null && !this.ac) {
                return this.onKeyDown(a);
            }
            return false;
        }
        af(a) {
            if (this.onKeyUp != null && !this.ac) {
                return this.onKeyUp(a);
            }
            return false;
        }
        aw(a) {
            if (this.onDoubleTap != null && !this.ac) {
                this.onDoubleTap(a);
            }
        }
        a1(a) {
            if (this.onHold != null && !this.ac) {
                this.onHold(a);
            }
        }
        get ac() {
            return this.z;
        }
        set ac(a) {
            this.z = a;
            this.as();
        }
        as() {
        }
    }
    EventProxy.$t = /*@__PURE__*/ markType(EventProxy, 'EventProxy');
    return EventProxy;
})();
