// igniteui-react-core/esm2015/lib/HorizontalStackedSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSeriesBaseDescription } from "./StackedSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HorizontalStackedSeriesBaseDescription = /*@__PURE__*/ (() => {
    class HorizontalStackedSeriesBaseDescription extends StackedSeriesBaseDescription {
        constructor() {
            super();
            this.d5 = null;
            this.d6 = null;
        }
        get_type() {
            return "HorizontalStackedSeriesBase";
        }
        get xAxisRef() {
            return this.d5;
        }
        set xAxisRef(a) {
            this.d5 = a;
            this.e("XAxisRef");
        }
        get yAxisRef() {
            return this.d6;
        }
        set yAxisRef(a) {
            this.d6 = a;
            this.e("YAxisRef");
        }
    }
    HorizontalStackedSeriesBaseDescription.$t = /*@__PURE__*/ markType(HorizontalStackedSeriesBaseDescription, 'HorizontalStackedSeriesBaseDescription', StackedSeriesBaseDescription.$);
    return HorizontalStackedSeriesBaseDescription;
})();
