// igniteui-react-core/esm2015/lib/TimeAxisBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAxisBaseDescription } from "./CategoryAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TimeAxisBaseDescription = /*@__PURE__*/ (() => {
    class TimeAxisBaseDescription extends CategoryAxisBaseDescription {
        constructor() {
            super();
            this.dz = null;
            this.dm = false;
            this.dq = new Date();
            this.dp = new Date();
            this.ds = new Date();
            this.dr = new Date();
            this.dy = null;
            this.dx = null;
        }
        get_type() {
            return "TimeAxisBase";
        }
        get dateTimeMemberPath() {
            return this.dz;
        }
        set dateTimeMemberPath(a) {
            this.dz = a;
            this.e("DateTimeMemberPath");
        }
        get isDataPreSorted() {
            return this.dm;
        }
        set isDataPreSorted(a) {
            this.dm = a;
            this.e("IsDataPreSorted");
        }
        get actualMinimumValue() {
            return this.dq;
        }
        set actualMinimumValue(a) {
            this.dq = a;
            this.e("ActualMinimumValue");
        }
        get actualMaximumValue() {
            return this.dp;
        }
        set actualMaximumValue(a) {
            this.dp = a;
            this.e("ActualMaximumValue");
        }
        get minimumValue() {
            return this.ds;
        }
        set minimumValue(a) {
            this.ds = a;
            this.e("MinimumValue");
        }
        get maximumValue() {
            return this.dr;
        }
        set maximumValue(a) {
            this.dr = a;
            this.e("MaximumValue");
        }
        get actualMinimumValueChangeRef() {
            return this.dy;
        }
        set actualMinimumValueChangeRef(a) {
            this.dy = a;
            this.e("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.dx;
        }
        set actualMaximumValueChangeRef(a) {
            this.dx = a;
            this.e("ActualMaximumValueChangeRef");
        }
    }
    TimeAxisBaseDescription.$t = /*@__PURE__*/ markType(TimeAxisBaseDescription, 'TimeAxisBaseDescription', CategoryAxisBaseDescription.$);
    return TimeAxisBaseDescription;
})();
