// igniteui-react-core/esm2015/lib/TimeXAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { TimeAxisBaseDescription } from "./TimeAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TimeXAxisDescription = /*@__PURE__*/ (() => {
    class TimeXAxisDescription extends TimeAxisBaseDescription {
        constructor() {
            super();
            this.d3 = null;
            this.d9 = null;
            this.d5 = null;
            this.d4 = null;
        }
        get_type() {
            return "TimeXAxis";
        }
        get breaks() {
            return this.d3;
        }
        set breaks(a) {
            this.d3 = a;
            this.e("Breaks");
        }
        get labellingMode() {
            return this.d9;
        }
        set labellingMode(a) {
            this.d9 = a;
            this.e("LabellingMode");
        }
        get labelFormats() {
            return this.d5;
        }
        set labelFormats(a) {
            this.d5 = a;
            this.e("LabelFormats");
        }
        get intervals() {
            return this.d4;
        }
        set intervals(a) {
            this.d4 = a;
            this.e("Intervals");
        }
    }
    TimeXAxisDescription.$t = /*@__PURE__*/ markType(TimeXAxisDescription, 'TimeXAxisDescription', TimeAxisBaseDescription.$);
    return TimeXAxisDescription;
})();
