// igniteui-react-core/esm2015/lib/WebButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebButtonBaseDescription } from "./WebButtonBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebButtonDescription = /*@__PURE__*/ (() => {
    class WebButtonDescription extends WebButtonBaseDescription {
        constructor() {
            super();
            this.ae = null;
            this.z = null;
            this.y = null;
            this.aa = null;
        }
        get_type() {
            return "WebButton";
        }
        get variant() {
            return this.ae;
        }
        set variant(a) {
            this.ae = a;
            this.e("Variant");
        }
        get prefix() {
            return this.z;
        }
        set prefix(a) {
            this.z = a;
            this.e("Prefix");
        }
        get content() {
            return this.y;
        }
        set content(a) {
            this.y = a;
            this.e("Content");
        }
        get suffix() {
            return this.aa;
        }
        set suffix(a) {
            this.aa = a;
            this.e("Suffix");
        }
    }
    WebButtonDescription.$t = /*@__PURE__*/ markType(WebButtonDescription, 'WebButtonDescription', WebButtonBaseDescription.$);
    return WebButtonDescription;
})();
