// igniteui-react-core/esm2015/lib/WebCheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebCheckboxBaseDescription } from "./WebCheckboxBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCheckboxDescription = /*@__PURE__*/ (() => {
    class WebCheckboxDescription extends WebCheckboxBaseDescription {
        constructor() {
            super();
            this.ae = false;
            this.ac = null;
        }
        get_type() {
            return "WebCheckbox";
        }
        get indeterminate() {
            return this.ae;
        }
        set indeterminate(a) {
            this.ae = a;
            this.e("Indeterminate");
        }
        get content() {
            return this.ac;
        }
        set content(a) {
            this.ac = a;
            this.e("Content");
        }
    }
    WebCheckboxDescription.$t = /*@__PURE__*/ markType(WebCheckboxDescription, 'WebCheckboxDescription', WebCheckboxBaseDescription.$);
    return WebCheckboxDescription;
})();
