// igniteui-react-core/esm2015/lib/WebCardHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCardHeaderDescription = /*@__PURE__*/ (() => {
    class WebCardHeaderDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.i = null;
            this.g = null;
            this.f = null;
        }
        get_type() {
            return "WebCardHeader";
        }
        get type() {
            return this.get_type();
        }
        get thumbnail() {
            return this.h;
        }
        set thumbnail(a) {
            this.h = a;
            this.e("Thumbnail");
        }
        get title() {
            return this.i;
        }
        set title(a) {
            this.i = a;
            this.e("Title");
        }
        get subtitle() {
            return this.g;
        }
        set subtitle(a) {
            this.g = a;
            this.e("Subtitle");
        }
        get content() {
            return this.f;
        }
        set content(a) {
            this.f = a;
            this.e("Content");
        }
    }
    WebCardHeaderDescription.$t = /*@__PURE__*/ markType(WebCardHeaderDescription, 'WebCardHeaderDescription', Description.$);
    return WebCardHeaderDescription;
})();
