// igniteui-react-core/esm2015/lib/WebListItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebListItemDescription = /*@__PURE__*/ (() => {
    class WebListItemDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.i = null;
            this.f = null;
            this.g = null;
        }
        get_type() {
            return "WebListItem";
        }
        get type() {
            return this.get_type();
        }
        get start() {
            return this.h;
        }
        set start(a) {
            this.h = a;
            this.e("Start");
        }
        get title() {
            return this.j;
        }
        set title(a) {
            this.j = a;
            this.e("Title");
        }
        get subtitle() {
            return this.i;
        }
        set subtitle(a) {
            this.i = a;
            this.e("Subtitle");
        }
        get content() {
            return this.f;
        }
        set content(a) {
            this.f = a;
            this.e("Content");
        }
        get end() {
            return this.g;
        }
        set end(a) {
            this.g = a;
            this.e("End");
        }
    }
    WebListItemDescription.$t = /*@__PURE__*/ markType(WebListItemDescription, 'WebListItemDescription', Description.$);
    return WebListItemDescription;
})();
