// igniteui-react-core/esm2015/lib/WebNavDrawerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebNavDrawerDescription = /*@__PURE__*/ (() => {
    class WebNavDrawerDescription extends Description {
        constructor() {
            super();
            this.p = null;
            this.j = false;
            this.f = null;
            this.g = null;
            this.o = null;
            this.n = null;
            this.m = null;
            this.l = null;
        }
        get_type() {
            return "WebNavDrawer";
        }
        get type() {
            return this.get_type();
        }
        get position() {
            return this.p;
        }
        set position(a) {
            this.p = a;
            this.e("Position");
        }
        get open() {
            return this.j;
        }
        set open(a) {
            this.j = a;
            this.e("Open");
        }
        get content() {
            return this.f;
        }
        set content(a) {
            this.f = a;
            this.e("Content");
        }
        get mini() {
            return this.g;
        }
        set mini(a) {
            this.g = a;
            this.e("Mini");
        }
        get openingRef() {
            return this.o;
        }
        set openingRef(a) {
            this.o = a;
            this.e("OpeningRef");
        }
        get openedRef() {
            return this.n;
        }
        set openedRef(a) {
            this.n = a;
            this.e("OpenedRef");
        }
        get closingRef() {
            return this.m;
        }
        set closingRef(a) {
            this.m = a;
            this.e("ClosingRef");
        }
        get closedRef() {
            return this.l;
        }
        set closedRef(a) {
            this.l = a;
            this.e("ClosedRef");
        }
    }
    WebNavDrawerDescription.$t = /*@__PURE__*/ markType(WebNavDrawerDescription, 'WebNavDrawerDescription', Description.$);
    return WebNavDrawerDescription;
})();
