// igniteui-react-core/esm2015/lib/WebNavDrawerItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebNavDrawerItemDescription = /*@__PURE__*/ (() => {
    class WebNavDrawerItemDescription extends Description {
        constructor() {
            super();
            this.k = false;
            this.j = false;
            this.g = null;
            this.f = null;
        }
        get_type() {
            return "WebNavDrawerItem";
        }
        get type() {
            return this.get_type();
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.e("Disabled");
        }
        get active() {
            return this.j;
        }
        set active(a) {
            this.j = a;
            this.e("Active");
        }
        get icon() {
            return this.g;
        }
        set icon(a) {
            this.g = a;
            this.e("Icon");
        }
        get content() {
            return this.f;
        }
        set content(a) {
            this.f = a;
            this.e("Content");
        }
    }
    WebNavDrawerItemDescription.$t = /*@__PURE__*/ markType(WebNavDrawerItemDescription, 'WebNavDrawerItemDescription', Description.$);
    return WebNavDrawerItemDescription;
})();
