// igniteui-react-charts/esm5/lib/CalloutAnnotation_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DependencyObject } from "igniteui-react-core";
import { INotifyPropertyChanged_$type, Base, PropertyChangedEventArgs, typeCast, markType, markDep } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { Series } from "./Series";
import { Brush } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
/**
 * @hidden
 */
var CalloutAnnotation = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutAnnotation, _super);
    function CalloutAnnotation() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.propertyChanged = null;
        _this.k = null;
        _this.ab = 0;
        _this.ac = 0;
        _this.am = null;
        _this.aw = null;
        _this.aa = null;
        _this.z = null;
        _this.l = null;
        _this._formatLabel = null;
        _this.bk = null;
        _this.bh = null;
        _this.bj = null;
        _this.bi = null;
        _this.x = 0;
        _this.o = false;
        _this.n = true;
        _this.bo = null;
        _this.u = NaN;
        _this.w = NaN;
        _this.v = NaN;
        _this.t = NaN;
        return _this;
    }
    CalloutAnnotation.prototype.a0 = function (a, b, c) {
        this.az(a, b, c);
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    CalloutAnnotation.prototype.az = function (a, b, c) {
        switch (a) {
            case CalloutAnnotation.$$p[13]:
                this.ab = c;
                break;
            case CalloutAnnotation.$$p[14]:
                this.ac = c;
                break;
            case CalloutAnnotation.$$p[9]:
                this.l = c;
                break;
            case CalloutAnnotation.$$p[5]:
                this.z = c;
                break;
            case CalloutAnnotation.$$p[6]:
                this.aa = c;
                break;
            case CalloutAnnotation.$$p[12]:
                this.am = c;
                break;
            case CalloutAnnotation.$$p[11]:
                this.bk = c;
                break;
            case CalloutAnnotation.$$p[7]:
                this.bi = c;
                break;
            case CalloutAnnotation.$$p[8]:
                this.bj = c;
                break;
            case CalloutAnnotation.$$p[4]:
                this.bh = c;
                break;
            case CalloutAnnotation.$$p[10]:
                this.x = c;
                break;
            case CalloutAnnotation.$$p[1]:
                this.n = true;
                this.u = c;
                break;
            case CalloutAnnotation.$$p[3]:
                this.n = true;
                this.w = c;
                break;
            case CalloutAnnotation.$$p[2]:
                this.n = true;
                this.v = c;
                break;
            case CalloutAnnotation.$$p[0]:
                this.n = true;
                this.t = c;
                break;
        }
        this.o = true;
    };
    Object.defineProperty(CalloutAnnotation.prototype, "formatLabel", {
        get: function () {
            return this._formatLabel;
        },
        set: function (a) {
            this._formatLabel = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotation.prototype, "backgroundPadding", {
        get: function () {
            if (this.n) {
                this.n = false;
                this.bo = new Thickness(1, this.u, this.w, this.v, this.t);
            }
            return this.bo;
        },
        set: function (a) {
            this.backgroundPaddingLeft = a.c;
            this.backgroundPaddingTop = a.e;
            this.backgroundPaddingRight = a.d;
            this.backgroundPaddingBottom = a.b;
        },
        enumerable: false,
        configurable: true
    });
    CalloutAnnotation.$t = markType(CalloutAnnotation, 'CalloutAnnotation', DependencyObject.$, [INotifyPropertyChanged_$type]);
    CalloutAnnotation.$$p = markDep(DependencyProperty, PropertyMetadata, CalloutAnnotation, 'a0', ['BackgroundPaddingBottom', [1, NaN], 'BackgroundPaddingLeft', [1, NaN], 'BackgroundPaddingRight', [1, NaN], 'BackgroundPaddingTop', [1, NaN], 'Background', [Brush.$, null], 'Content', [4, null], 'Key', [4, null], 'LeaderBrush', [Brush.$, null], 'Outline', [Brush.$, null], 'Series', [Series.$, null], 'StrokeThickness', [1, 0], 'TextColor', [Brush.$, null], 'Text', [2, null], 'XValue', [4, 0], 'YValue', [4, 0]]);
    return CalloutAnnotation;
}(DependencyObject));
export { CalloutAnnotation };
/**
 * @hidden
 */
var CalloutAnnotationWrapper = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutAnnotationWrapper, _super);
    function CalloutAnnotationWrapper() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = null;
        _this.c = null;
        return _this;
    }
    Object.defineProperty(CalloutAnnotationWrapper.prototype, "b", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            this.a = a;
            this.a.k = this;
        },
        enumerable: false,
        configurable: true
    });
    CalloutAnnotationWrapper.prototype.equals = function (a) {
        var b = typeCast(CalloutAnnotationWrapper.$, a);
        if (b == null) {
            return _super.prototype.equals.call(this, a);
        }
        if (b.b == null && this.b == null) {
            return _super.prototype.equals.call(this, a);
        }
        if (b.b == null) {
            return false;
        }
        if (this.b.aa != null || b.b.aa != null) {
            return Base.equalsStatic(this.b.aa, b.b.aa);
        }
        return _super.prototype.equals.call(this, a);
    };
    CalloutAnnotationWrapper.prototype.getHashCode = function () {
        if (this.b == null || this.b.aa == null) {
            return _super.prototype.getHashCode.call(this);
        }
        return Base.getHashCodeStatic(this.b.aa);
    };
    CalloutAnnotationWrapper.$t = markType(CalloutAnnotationWrapper, 'CalloutAnnotationWrapper');
    return CalloutAnnotationWrapper;
}(Base));
export { CalloutAnnotationWrapper };
